src/ua/main.cpp

```diff
@@ -529,6 +543,39 @@ void printEnvironmentInfo() {
   }
 }
 
+// There is currently the possibility of a race condition if a chunk
+// upload timed-out.  It's possible that a second upload succeeds,
+// has the chunk marked as "complete" and then the first request makes
+// its way through the queue and marks the chunk as pending again.
+// Since we are just about to close the file, we'll check to see if any
+// chunks are marked as pending, and if so, we'll retry them.
+void check_for_complete_chunks(vector<File> &files) {
+  for (int currCheckNum=0; currCheckNum < NUM_CHUNK_CHECKS; ++currCheckNum){
+    while (!chunksFinished.empty()) {
+      Chunk *c = chunksFinished.consume();
+      if (!is_chunk_complete(c)) {
+          chunksToUpload.produce(c);
+      }
+    }
+    // All of the chunks were marked as complete, so let's exit and we
+    // should be safeish to close the file.
+    if(chunksToUpload.size() == 0)
```

Comment:
I don't think so.  The way the code is currently written is let's say we check all chunks and they are all marked as complete (which should happen 99.999% of the time).  Then this break will exit us out of this retry immediately (not slowing anything down from current behavior).

In the rare case we have a chunk that needs to be uploaded, we'll blow by this check and then the next little clip of code will actually push the chunks to be uploaded back on the queue of items to upload, and then the following little clip of code will do the pause to wait for chunksToUpload to drain.  Then we'll re-enter this check loop, check to see if all these new chunks that had to be reuploaded are now marked complete etc.